searx/doc.py

```diff
@@ -0,0 +1,290 @@
+    # found ..
+    import time
+
+    time.sleep(3)
+    return ret_val
+
+
+class HTMLPage(typing.NamedTuple):
+    """HTML page for online documentations."""
+
+    title: str
+    html: str
+
+
+class Markup:
+    """Base class to render (online) documentation.
+
+    The base class just render the jinja template, therefore the base class does
+    not need to have a special source markup.  To e.g. render Markdown
+    :py:class:`online HTML <HTMLPage>` page , see :py:obj:`Mistletoe`.
+
+    """
+
+    def __init__(self, suffix='.md'):
+        self.suffix = suffix
+
+    def render(self, content: str) -> str:
+        """Render Jinja content"""
+        ctx = self.get_ctx()
+        template = jinja2.Environment().from_string(content)
+        return template.render(**ctx)
+
+    @classmethod
+    def get_ctx(cls):
+        """Common Jinja context."""
+
+        ctx = {}
+        ctx['GIT_URL'] = GIT_URL
+        ctx['get_setting'] = get_setting
+
+        # markup of the content is unspecified in this base class, just insert
+        # link names (first argument).
+
+        ctx['link'] = lambda *args, **kwargs: args[0]
+        ctx['search'] = lambda *args, **kwargs: args[0]
+
+        return ctx
+
+
+class Markdown(Markup):
+    """Markdown (CommonMark_) Jinja template.
+
+    .. _CommonMark: https://commonmark.org/
+    """
+
+    @classmethod
+    def get_ctx(cls):
+        def _md_link(name, url):
+            return "[%s](%s)" % (name, url_for(url))
+
+        def _md_search(query):
+            url = '%s?q=%s' % (url_for('search'), urllib.parse.quote(query))
+            return '[%s](%s)' % (query, url)
+
+        ctx = Markup.get_ctx()
+        ctx['link'] = _md_link
+        ctx['search'] = _md_search
+
+        return ctx
+
+
+class Mistletoe(Markdown):
+    """Render :py:class:`Markdown` to :py:class:`online HTML <HTMLPage>` by using
+     mistletoe_.
+
+    .. _mistletoe: https://github.com/miyuchina/mistletoe
+
+    """
+
+    def render(self, content: str) -> HTMLPage:
+        title = ''
+        markdown = []
+        for l in content.split('\n'):
+            if not title and l.startswith('# '):
+                title = l.strip('# ')
+            else:
+                markdown.append(l)
+        markdown = '\n'.join(markdown)
+
+        markdown = super().render(markdown)
+        html = mistletoe.markdown(markdown)
+        return HTMLPage(title=title, html=html)
+
+
+class Offline(Markdown):
+    """Render online documentation for offline Sphinx-doc (MyST)_.
+
+    .. _MyST: https://www.sphinx-doc.org/en/master/usage/markdown.html
+
+    """
+
+    @classmethod
+    def get_ctx(cls):
+
+        ctx = Markup.get_ctx()
+        ctx['link'] = lambda *args, **kwargs: '`%s`' % args[0]
+        ctx['search'] = lambda *args, **kwargs: '`%s`' % args[0]
+
+        return ctx
+
+
+_DOC_FOLDER = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'doc')
```

Comment:
Is there a problem to use the `pkg_resources` instead of the `os` package and `open`?

Reply:
My intention is different.  IMO (not in this PR but may be in the future) the location is taken from settings.yml / see https://github.com/searxng/searxng/pull/930/files#diff-b439f9f6bc445fcbb1be1c7d0314a845126d56ba235e52361c9918d943d5733aR252-R258

```
        # future: could be set from settings.xml
        self.folder: str = _DOC_FOLDER
        self.i18n_origin: str = 'en'
        self.l10n: typing.List = [
            'en',
        ]
        self.toc: typing.List = ['search-syntax', 'about']
```

By this instance maintainer can write their own online documentation.  By example in my instance I use *hostname replace* to link all youtube results to invidious instances .. and I want to document this setup to my users.

Reply:
>  I use hostname replace to link all youtube results to invidious instances

This need exists for everyone using this plugin.
IMO, the plugin should provide a documentation or at least some information in the `/preferences` page (there is no way to do the later currently)

Reply:
This plugin was just an example .. I think there is a need for having a online documentation to describe the intention of the setup of that instance.

By example: [... but all results (incl. yahoo) are redirected to an instance in DE where the majority of my users is located](https://github.com/searxng/searxng/discussions/911)

Reply:
> By example: https://github.com/searxng/searxng/discussions/911

Where this documentation should be?
Currently it requires to change `searx/doc.py` to include the additional pages in the tab.

Idea:
* keep the git repository clean
* allow to add additional documentation repository: all pages declared here are included in the tab.

So the administrator just have to declare the directory, add some files nothing more.

(perhaps it could be better, simpler)

Reply:
> So the administrator just have to declare the directory, add some files nothing more.

Sounds good to me .. 

In this PR I want to prepare a boilerplate that is able to do so, but I would like to implement the functionality (where the values are taken from settings.yml) in another PR .. is it OK for you?

Reply:
IMO, let's do that in another PR and keep the focus on the rendering of the `searx/doc` content.